lib/datadog/di/instrumenter.rb

```diff
@@ -0,0 +1,301 @@
+    #
+    # Note that the upstream code is responsible for generally storing Probes.
+    # This is normally accomplished by ProbeManager. ProbeManager stores all
+    # known probes, instrumented or not, and is responsible for calling
+    # +unhook+ of Instrumenter to clean up instrumentation when a user
+    # deletes a probe in UI or when DI is shut down.
+    #
+    # Given the need to store state, and also that there are several Probe
+    # attributes that affect how instrumentation is set up and that must be
+    # consulted very early in the callback invocation (e.g., to perform
+    # rate limiting correctly), Instrumenter takes Probe instances as
+    # arguments rather than e.g. file + line number or class + method name.
+    # As a result, Instrumenter is rather coupled to DI the product and is
+    # not trivially usable as a general-purpose Ruby instrumentation tool
+    # (however, Probe instances can be replaced by OpenStruct instances
+    # providing the same interface with not much effort).
+    #
+    # @api private
+    class Instrumenter
+      def initialize(settings, serializer, logger, code_tracker: nil)
+        @settings = settings
+        @serializer = serializer
+        @logger = logger
+        @code_tracker = code_tracker
+
+        @lock = Mutex.new
+      end
+
+      attr_reader :settings
+      attr_reader :serializer
+      attr_reader :logger
+      attr_reader :code_tracker
+
+      # This is a substitute for Thread::Backtrace::Location
+      # which does not have a public constructor.
+      # Used for the fabricated stack frame for the method itself
+      # for method probes (which use Module#prepend and thus aren't called
+      # from the method but from outside of the method).
+      Location = Struct.new(:path, :lineno, :label)
+
+      def hook_method(probe, &block)
+        unless block
+          raise ArgumentError, 'block is required'
+        end
+
+        lock.synchronize do
+          if probe.instrumentation_module
+            # Already instrumented, warn?
+            return
+          end
+        end
+
+        cls = symbolize_class_name(probe.type_name)
+        serializer = self.serializer
+        method_name = probe.method_name
+        target_method = cls.instance_method(method_name)
+        loc = target_method.source_location
+        rate_limiter = probe.rate_limiter
+
+        mod = Module.new do
+          define_method(method_name) do |*args, **kwargs| # steep:ignore
+            if rate_limiter.nil? || rate_limiter.allow?
+              # Arguments may be mutated by the method, therefore
+              # they need to be serialized prior to method invocation.
+              entry_args = if probe.capture_snapshot?
+                serializer.serialize_args(args, kwargs)
+              end
+              rv = nil
+              duration = Benchmark.realtime do # steep:ignore
+                rv = super(*args, **kwargs)
+              end
```

Comment:
I'm not sure this is delegation is correct -- there may be corner cases with args/keyword args; see https://eregon.me/blog/2019/11/10/the-delegation-challenge-of-ruby27.html for details. I think this is a quite sensitive part since getting it wrong means installing a probe can break the app.

(Also @lloeki quite likes this topic so I think he's the person to ping for a quick check on this bit ;D )

Reply:
This is one of my items to address before initial release, but the "shortcut" is I can just mandate Ruby 3.0+ for DI if this problem is difficult to solve.